src/core/annotation.js

```diff
@@ -929,17 +984,22 @@ class TextAnnotation extends MarkupAnnotation {
 
     super(parameters);
 
+    const dict = parameters.dict;
     this.data.annotationType = AnnotationType.TEXT;
 
     if (this.data.hasAppearance) {
       this.data.name = 'NoIcon';
     } else {
       this.data.rect[1] = this.data.rect[3] - DEFAULT_ICON_SIZE;
       this.data.rect[2] = this.data.rect[0] + DEFAULT_ICON_SIZE;
-      this.data.name = parameters.dict.has('Name') ?
-                       parameters.dict.get('Name').name : 'Note';
+      this.data.name = dict.has('Name') ?
+                       dict.get('Name').name : 'Note';
     }
 
+    if (dict.has('State')) {
```

Comment:
The spec states: "The state is not specified in the annotation itself but in a separate text annotation that refers to the original annotation by means of its IRT (“in reply to”) entry."

Should we check if an `IRT` entry is available here? Moreover, should this text annotation be rendered (which I think is the case at the moment) or should those annotations not be rendered because they only contain the state of another annotation? I think it's rather odd that the spec uses a different annotation for keeping state of another annotation, let alone it being a `Text` annotation, but that seems to be how it works...

Whether or not an annotation should be rendered is controlled in https://github.com/mozilla/pdf.js/blob/master/src/display/annotation_layer.js#L358-L359, so you could add a check for `parameters.data.state !== null` there.

Finally, I think the data attributes for an annotation type should always be provided and set to `null` if they are not used. This way the API is consistent: you always get all attributes on which you can do a check. In this case, the `state` and `stateModel` properties would alway be provided, but set to `null` if they don't have a value.

Reply:
Yeah this is tricky … Maybe, as there is viewable and printable in Annotation base class, there should also exist renderable, which will tell, wheter annotation should be rendered into canvas ([added into opList (see code)](https://github.com/mozilla/pdf.js/blob/master/src/core/document.js#L243) ). Viewable will then tell, wheter it can be passed out of worker at all ([added into getAnnotationsData(see code)](https://github.com/mozilla/pdf.js/blob/master/src/core/document.js#L299) response). 

annotation.viewable is now used in both of those cases, but IMO it should be splitted. So in core/document.js there should exist isAnnotationRenderable as is now (which will check for annotation.renderable - for opList) and isAnnotationViewable (which will check for annotation.viewable - for getAnnotationsData). 

RT=R and State annotations would be renderable=false, but viewable=true.  Standard annotations and Group annotations (probably?) would be true,true. And those, which are now viewable=false would be false,false. But I would rather keep that into some future PRs as it might break a thing or two ...

Reply:
That sounds like a good approach, but I agree it's best to keep that for a follow-up PR given its complexity.